src/daemon/plugin.c

```diff
@@ -1085,7 +1085,48 @@ static int plugin_insert_read(read_func_t *rf) {
   int status;
   llentry_t *le;
 
-  rf->rf_next_read = cdtime();
+  /* set the time of the first read according to "StartRead" in collectd.conf
+     start minute, second, millisecond is given as float in seconds */
+  double start = rf->rf_ctx.start_time; /* (default is -1.0) */
+  if (start < 0) {
+    /* if StartAt is not defined, start reading immediately */
+    rf->rf_next_read = cdtime();
+  } else {
+    /* get minute in hour and second+millisecond in minute */
```

Comment:
IMHO: this code is overcomplicated. All you need is `rf->rf_next_read = cdtime() + rf->rf_ctx.first_read_delay; `

Default should be 0.

Reply:
Do you mean the code that determines the `first_read_delay` is overcomplicated?
The if-branch takes care of the default case, when _StartRead_ has not been set. Thus, it executes `rf->rf_next_read = cdtime()` as done before. The else-branch determines when the next read should happen. This depends on the current time `cdtime()`, _StartRead_ and _Interval_ values.

Reply:
> All you need is `rf->rf_next_read = cdtime() + rf->rf_ctx.first_read_delay; `

Determine the first_read_delay is the "complicated" part.